extensions/harvester/sakubeconfig.go

```diff
@@ -0,0 +1,93 @@
+package harvester
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	provv1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
+	"github.com/rancher/shepherd/clients/rancher"
+	steveV1 "github.com/rancher/shepherd/clients/rancher/v1"
+	"github.com/rancher/shepherd/extensions/clusters"
+)
+
+const (
+	HarvesterProviderClusterLabel = "labelSelector=provider.cattle.io=harvester"
+)
+
+// KubeConfigOpts is configuration need to get a kubeconfig with SA
+type KubeConfigOpts struct {
+	CSIClusterRoleName string `yaml:"csiclusterRoleName" json:"csiclusterRoleName" default:"harvesterhci.io:csi-driver"`
+	ClusterRoleName    string `yaml:"clusterRoleName" json:"clusterRoleName" default:"harvesterhci.io:cloudprovider"`
+	Namespace          string `yaml:"namespace" json:"namespace" default:"default"`
+	SaName             string `yaml:"serviceAccountName" json:"serviceAccountName" default:"rancherCluster"`
+}
+
+// GetHarvesterSAKubeconfig generates a kubeconfig from the harvester cluster with a custom SA for the
+// cluster which will use the kubeconfig. This is typically used when creating a downstream cluster for
+// harvester that uses the harvester cloud provider.
+func GetHarvesterSAKubeconfig(client *rancher.Client, clusterName string) ([]byte, error) {
+	query, err := url.ParseQuery(HarvesterProviderClusterLabel)
+	if err != nil {
+		return nil, err
+	}
+
+	harvesterCluster, err := client.Steve.SteveType(clusters.ProvisioningSteveResourceType).ListAll(query)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(harvesterCluster.Data) == 0 {
+		return nil, errors.New("no imported harvester cluster found")
+	}
+
+	status := &provv1.ClusterStatus{}
+	err = steveV1.ConvertToK8sType(harvesterCluster.Data[0].Status, status)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := KubeConfigOpts{
+		CSIClusterRoleName: "harvesterhci.io:csi-driver",
+		ClusterRoleName:    "harvesterhci.io:cloudprovider",
+		Namespace:          "default",
+		SaName:             clusterName,
+	}
+
+	bodyContent, err := json.Marshal(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("https://%s/k8s/clusters/%s/v1/harvester/kubeconfig", client.RancherConfig.Host, status.ClusterName), bytes.NewBuffer(bodyContent))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Authorization", "Bearer "+client.RancherConfig.AdminToken)
```

Comment:
also might need to change this. we have admin token here. if the client is not admin, the request might fail.

Reply:
Since, the adminClient was removed from the current version, the client we use will be a non admin client. In that case a non admin client would be using the admin token..

Reply:
So Caleb if we're keeping it accessible to standard non-admin clients then the clients bearer token should be used like this: 
client.Management.APIBaseClient.Opts.TokenKey try this out